src/Orleans.Core/Networking/Connection.cs

```diff
@@ -105,13 +105,13 @@ private static Task OnConnectedAsync(ConnectionContext context)
             return connection.RunInternal();
         }
 
-        protected virtual async Task RunInternal()
+        protected virtual Task RunInternal()
         {
             _transport = this.Context.Transport;
             _processIncomingTask = this.ProcessIncoming();
             _processOutgoingTask = this.ProcessOutgoing();
             _initializationTcs.TrySetResult(0);
-            await Task.WhenAll(_processIncomingTask, _processOutgoingTask);
+            return Task.WhenAll(_processIncomingTask, _processOutgoingTask);
```

Comment:
This was written this way for debugging purposes. There are generally only 10s-100s of connections, and they are each very long-lived, so per-connection state machine allocations are ok.